test/zdtm/static/sock_tcp_opts00.c

```diff
@@ -0,0 +1,97 @@
+#include <sys/socket.h>
+#include <sys/types.h>
+#include <netinet/in.h>
+#include <netinet/tcp.h>
+
+#include "zdtmtst.h"
+
+const char *test_doc = "Check that different tcp socket options are restored";
+const char *test_author = "Juntong Deng <juntong.deng@outlook.com>";
+
+#ifdef ZDTM_VAL_ZERO
+#define TCP_OPT_VAL 0
+#else
+#define TCP_OPT_VAL 1
+#endif
+
+#ifndef SOL_TCP
+#define SOL_TCP 6
+#endif
+
+struct sk_opt {
+	int level;
+	int opt;
+	int val;
+};
+
+struct sk_opt tcp_sk_opts[] = {
+	{ SOL_TCP, TCP_CORK, TCP_OPT_VAL },
+	{ SOL_TCP, TCP_NODELAY, TCP_OPT_VAL },
+	{ SOL_TCP, TCP_QUICKACK, TCP_OPT_VAL },
+};
+
+struct sk_conf {
+	int domain;
+	int type;
+	int protocol;
+	int sk;
+} sk_confs[] = {
+	{ AF_INET, SOCK_STREAM, IPPROTO_TCP },
+	{ AF_INET6, SOCK_STREAM, IPPROTO_TCP },
+};
+
+int main(int argc, char **argv)
+{
+	struct sk_opt *opts = tcp_sk_opts;
+	int n_opts = ARRAY_SIZE(tcp_sk_opts);
+	int exit_code = 1;
+	int i, j, val;
+	socklen_t len;
+
+	test_init(argc, argv);
+
+	for (i = 0; i < ARRAY_SIZE(sk_confs); i++) {
+		sk_confs[i].sk = socket(sk_confs[i].domain, sk_confs[i].type, sk_confs[i].protocol);
+		if (sk_confs[i].sk == -1) {
+			pr_perror("socket(%d,%d,%d) failed", sk_confs[i].domain, sk_confs[i].type,
+				  sk_confs[i].protocol);
+			goto close;
+		}
+	}
+
+	for (i = 0; i < ARRAY_SIZE(sk_confs); i++) {
+		for (j = 0; j < n_opts; j++) {
+			val = opts[j].val;
+			if (setsockopt(sk_confs[i].sk, opts[j].level, opts[j].opt, &val, sizeof(int)) == -1) {
+				pr_perror("setsockopt(%d, %d) failed", opts[j].level, opts[j].opt);
+				goto close;
+			}
+		}
+	}
+
+	test_daemon();
+	test_waitsig();
+
+	for (i = 0; i < ARRAY_SIZE(sk_confs); i++) {
+		for (j = 0; j < n_opts; j++) {
+			len = sizeof(int);
+			if (getsockopt(sk_confs[i].sk, opts[j].level, opts[j].opt, &val, &len) == -1) {
+				pr_perror("getsockopt(%d, %d) failed", opts[j].level, opts[j].opt);
+				goto close;
+			}
+
+			if (val != opts[j].val) {
+				fail("Unexpected value socket(%d,%d,%d) opts(%d,%d)", sk_confs[i].domain,
+				     sk_confs[i].type, sk_confs[i].protocol, opts[j].level, opts[j].opt);
```

Comment:
> The test is failing with the following error in CI:
> 
> > FAIL: sock_tcp_opts00.c:84: Unexpected value socket(2,1,6) opts(6,3) (errno = 11 (Resource temporarily unavailable))

This error is because, in the current CRIU implementation, TCP_CORK, TCP_NODELAY, TCP_QUICKACK socket options are dumped only after the TCP socket establish connection. In the test case I added, it did not make the test socket establish connection, causing the error to occur (but strangely this test case worked when I used zdtm, which caused me to not notice the problem).

I think the current CRIU implementation is wrong because the application can set TCP socket options before TCP connection is established, so I think we should move the part of dump_tcp_conn_state() that dump TCP socket options to dump_one_tcp().

What do you think?

Reply:
1) I agree it looks like an error in tcp socket options dumping/restoring code that your test found. Options just disappear on non-connected sockets.

2) I don't see any problem with dumping these tcp socket options separately from connections similar to what we do for ip options here https://github.com/checkpoint-restore/criu/commit/c9a61a205f TcpStreamEntry is just not the right place for those options.